File: midonet-api/src/main/java/org/midonet/api/l4lb/rest_api/HealthMonitorResource.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.api.l4lb.rest_api;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import javax.annotation.security.RolesAllowed;
import javax.validation.Validator;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import com.google.inject.Inject;
import com.google.inject.servlet.RequestScoped;

import org.midonet.api.ResourceUriBuilder;
import org.midonet.api.rest_api.AbstractResource;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.cluster.DataClient;
import org.midonet.cluster.auth.AuthRole;
import org.midonet.cluster.rest_api.ConflictHttpException;
import org.midonet.cluster.rest_api.ResourceUris;
import org.midonet.cluster.rest_api.ServiceUnavailableHttpException;
import org.midonet.cluster.rest_api.VendorMediaType;
import org.midonet.cluster.rest_api.conversion.HealthMonitorDataConverter;
import org.midonet.cluster.rest_api.conversion.PoolDataConverter;
import org.midonet.cluster.rest_api.models.HealthMonitor;
import org.midonet.cluster.rest_api.models.Pool;
import org.midonet.event.topology.HealthMonitorEvent;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.NoStatePathException;
import org.midonet.midolman.state.StateAccessException;
import org.midonet.midolman.state.StatePathExistsException;
import org.midonet.midolman.state.l4lb.MappingStatusException;

import static org.midonet.cluster.rest_api.conversion.HealthMonitorDataConverter.fromData;
import static org.midonet.cluster.rest_api.conversion.HealthMonitorDataConverter.toData;
import static org.midonet.cluster.rest_api.validation.MessageProperty.MAPPING_STATUS_IS_PENDING;
import static org.midonet.cluster.rest_api.validation.MessageProperty.RESOURCE_EXISTS;
import static org.midonet.cluster.rest_api.validation.MessageProperty.getMessage;


@RequestScoped
public class HealthMonitorResource extends AbstractResource {

    private final static HealthMonitorEvent healthMonitorEvent
            = new HealthMonitorEvent();

    @Inject
    public HealthMonitorResource(RestApiConfig config, UriInfo uriInfo,
                                 SecurityContext context,
                                 DataClient dataClient, Validator validator) {
        super(config, uriInfo, context, dataClient, validator);
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Produces({ VendorMediaType.APPLICATION_HEALTH_MONITOR_COLLECTION_JSON,
            MediaType.APPLICATION_JSON })
    public List<HealthMonitor> list()
        throws StateAccessException, SerializationException,
               IllegalAccessException {

        List<org.midonet.cluster.data.l4lb.HealthMonitor> hmonitors;

        hmonitors = dataClient.healthMonitorsGetAll();
        List<HealthMonitor> healthMonitors = new ArrayList<>();
        if (hmonitors != null) {
            for (org.midonet.cluster.data.l4lb.HealthMonitor hmData : hmonitors) {
                healthMonitors.add(fromData(hmData, getBaseUri()));
            }
        }
        return healthMonitors;
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Path("{id}")
    @Produces({ VendorMediaType.APPLICATION_HEALTH_MONITOR_JSON,
            MediaType.APPLICATION_JSON })
    public HealthMonitor get(@PathParam("id") UUID id)
            throws StateAccessException, SerializationException,
                   IllegalAccessException {

        org.midonet.cluster.data.l4lb.HealthMonitor healthMonitorData =
            dataClient.healthMonitorGet(id);
        if (healthMonitorData == null) {
            throw notFoundException(id, ResourceUris.HEALTH_MONITORS);
        }

        return HealthMonitorDataConverter.fromData(healthMonitorData,
                                                   getBaseUri());
    }

    @DELETE
    @RolesAllowed({ AuthRole.ADMIN, AuthRole.TENANT_ADMIN })
    @Path("{id}")
    public void delete(@PathParam("id") UUID id)
            throws StateAccessException, SerializationException {

        try {
            dataClient.healthMonitorDelete(id);
            healthMonitorEvent.delete(id);
        } catch (NoStatePathException ex) {
            // Delete is idempotent, so just ignore.
        } catch (MappingStatusException ex) {
            throw new ServiceUnavailableHttpException(
                getMessage(MAPPING_STATUS_IS_PENDING, ex.getMessage()));
        }
    }

    @POST
    @RolesAllowed({ AuthRole.ADMIN, AuthRole.TENANT_ADMIN })
    @Consumes({ VendorMediaType.APPLICATION_HEALTH_MONITOR_JSON,
                MediaType.APPLICATION_JSON })
    public Response create(HealthMonitor hm)
        throws StateAccessException, SerializationException {

        validate(hm);

        try {
            UUID id = dataClient.healthMonitorCreate(toData(hm));
            healthMonitorEvent.create(id, dataClient.healthMonitorGet(id));
            return Response.created(
                    ResourceUriBuilder.getHealthMonitor(getBaseUri(), id))
                    .build();
        } catch (StatePathExistsException ex) {
            throw new ConflictHttpException(ex,
                    getMessage(RESOURCE_EXISTS, "health monitor", hm.id));
        }
    }

    @PUT
    @RolesAllowed({ AuthRole.ADMIN, AuthRole.TENANT_ADMIN })
    @Path("{id}")
    @Consumes({ VendorMediaType.APPLICATION_HEALTH_MONITOR_JSON,
                MediaType.APPLICATION_JSON })
    public void update(@PathParam("id") UUID id, HealthMonitor healthMonitor)
        throws StateAccessException, SerializationException {

        healthMonitor.id = id;
        validate(healthMonitor);

        try {
            dataClient.healthMonitorUpdate(toData(healthMonitor));
            healthMonitorEvent.update(id, dataClient.healthMonitorGet(id));
        } catch (NoStatePathException ex) {
            throw notFoundException(id, "health monitor");
        } catch (MappingStatusException ex) {
            throw new ServiceUnavailableHttpException(
                getMessage(MAPPING_STATUS_IS_PENDING, ex.getMessage()));
        }
    }

    @GET
    @RolesAllowed({ AuthRole.ADMIN })
    @Path("{id}" + ResourceUriBuilder.POOLS)
    @Produces({VendorMediaType.APPLICATION_POOL_COLLECTION_JSON,
               MediaType.APPLICATION_JSON})
    public List<Pool> listPools(@PathParam("id") UUID id)
        throws StateAccessException, SerializationException,
               IllegalAccessException {

        List<org.midonet.cluster.data.l4lb.Pool> dataPools;
        try {
            dataPools = dataClient.healthMonitorGetPools(id);
        } catch (NoStatePathException ex) {
            throw notFoundException(id, "health monitor");
        }

        List<Pool> pools = new ArrayList<>(dataPools.size());
        for (org.midonet.cluster.data.l4lb.Pool dataPool : dataPools) {
            pools.add(PoolDataConverter.fromData(dataPool, getBaseUri()));
        }

        return pools;
    }
}

File: cluster/midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/HealthMonitorResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{List => JList}
import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status

import scala.collection.JavaConverters._

import com.google.common.base.Objects
import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.annotation._
import org.midonet.cluster.rest_api.models.{Pool, HealthMonitor}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{NoOps, Ops, ResourceContext}

@ApiResource(version = 1)
@Path("health_monitors")
@RequestScoped
@AllowGet(Array(APPLICATION_HEALTH_MONITOR_JSON,
                APPLICATION_JSON))
@AllowList(Array(APPLICATION_HEALTH_MONITOR_COLLECTION_JSON,
                 APPLICATION_JSON))
@AllowCreate(Array(APPLICATION_HEALTH_MONITOR_JSON,
                   APPLICATION_JSON))
@AllowUpdate(Array(APPLICATION_HEALTH_MONITOR_JSON,
                   APPLICATION_JSON))
@AllowDelete
class HealthMonitorResource @Inject()(resContext: ResourceContext)
    extends MidonetResource[HealthMonitor](resContext) {

    @GET
    @Path("{id}")
    @Produces(Array(APPLICATION_HEALTH_MONITOR_JSON, APPLICATION_JSON))
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String):HealthMonitor = {
        val hm = getResource(classOf[HealthMonitor], id).getOrThrow
        hm
    }

    @GET
    @Produces(Array(APPLICATION_HEALTH_MONITOR_COLLECTION_JSON, APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[HealthMonitor] = {
        val hms = listResources(classOf[HealthMonitor]).getOrThrow
        hms.asJava
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = {
        try {
            val response = super.delete(id)
            if (response.getStatus == Status.NOT_FOUND.getStatusCode)
                MidonetResource.OkNoContentResponse
            else
                response
        } catch {
            case e: WebApplicationException
                if e.getResponse.getStatus == Status.NOT_FOUND.getStatusCode =>
                MidonetResource.OkNoContentResponse
        }
    }

    protected override def createFilter(healthMonitor: HealthMonitor): Ops = {
        healthMonitor.create()
        NoOps
    }

    protected override def updateFilter(to: HealthMonitor,
                                        from: HealthMonitor): Ops = {
        to.update(from)
        NoOps
    }

    @GET
    @Path("{id}/pools")
    @Produces(Array(APPLICATION_POOL_COLLECTION_JSON, APPLICATION_JSON))
    def listPools(@HeaderParam("Accept") accept: String,
                  @PathParam("id") id: String): JList[Pool] = {
        val pools = listResources(classOf[Pool]).getOrThrow
        val filteredPools = pools filter (
            p => Objects.equal(p.healthMonitorId.toString, id))
        filteredPools.asJava
    }
}